training_utils.py

```diff
@@ -0,0 +1,63 @@
+import torch
+from typing import Any, Callable, Dict
+
+_ATTR_REGISTRY: Dict[str, str] = {}
+_FUNC_REGISTRY: Dict[str, Callable[[Any], torch.Tensor]] = {}
+
+def _register_attr(cls_name: str, attr: str) -> None:
+    _ATTR_REGISTRY[cls_name] = attr
+
+def _register_func(cls_name: str, fn: Callable[[Any], torch.Tensor]) -> None:
+    _FUNC_REGISTRY[cls_name] = fn
+
+def unpack_forward_output(output: Any) -> torch.Tensor:
+    if isinstance(output, (tuple, list, dict)):
+        raise ValueError(
+            "Tuples/lists/dicts are not handled. "
+            "Add a model-specific wrapper or explicit registry rule."
+        )
+
+    if isinstance(output, torch.Tensor):
+        return output
+
+    cls_name = output.__class__.__name__
+
+    if cls_name in _ATTR_REGISTRY:
+        v = getattr(output, _ATTR_REGISTRY[cls_name], None)
+        if isinstance(v, torch.Tensor):
+            return v
+        raise ValueError(f"{cls_name}.{_ATTR_REGISTRY[cls_name]} is not a torch.Tensor")
+
+    if cls_name in _FUNC_REGISTRY:
+        v = _FUNC_REGISTRY[cls_name](output)
+        if isinstance(v, torch.Tensor):
+            return v
+        raise ValueError(f"Custom func for {cls_name} did not return a torch.Tensor")
```

Comment:
nit: Let's keep error messages as proper sentences.
```suggestion
        raise ValueError(f"Custom function for {cls_name} did not return a torch.Tensor.")
```